sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,513 @@
+	}
+
+	if errCode := resp.GetSupportedAttKeyIDNumRes.GetErrorCode(); errCode != 0 {
+		return 0, fmt.Errorf("failed to call getSupportedAttKeyIDNumRequest (error code = %d)", errCode)
+	}
+
+	akIdNum := resp.GetSupportedAttKeyIDNumRes.GetAttKeyIdNum()
+	if akIdNum == 0 {
+		return 0, fmt.Errorf("failed to call getSupportedAttKeyIDNumRequest: invalid Attestation Key Identity Number = %d", akIdNum)
+	}
+
+	return akIdNum, nil
+}
+
+func getSupportedAttKeyIDsRequest(num uint32) ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.GetSupportedAttKeyIDsReq = &pb.AesmServiceRequest_GetSupportedAttKeyIDsRequest{
+		BufSize: num * attestationKeyIdentityLength,
+		Timeout: aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetSupportedAttKeyIDsRes = &pb.AesmServiceResponse_GetSupportedAttKeyIDsResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, err
+	}
+
+	if errCode := resp.GetSupportedAttKeyIDsRes.GetErrorCode(); errCode != 0 {
+		return nil, fmt.Errorf("failed to call getSupportedAttKeyIDsRequest (error code = %d)", errCode)
+	}
+
+	akIdList := resp.GetSupportedAttKeyIDsRes.GetAttKeyIds()
+	if akIdList == nil {
+		return nil, fmt.Errorf("failed to call getSupportedAttKeyIDsRequest")
+	}
+
+	return akIdList, nil
+}
+
+func getAttestationKeyIdentity(quoteType string) ([]byte, error) {
+	attKeyId, err := selectAttKeyIDRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	akId := &attestationKeyIdentity{}
+	if err = restruct.Unpack(attKeyId, binary.LittleEndian, &akId); err != nil {
+		return nil, err
+	}
+
+	if strings.EqualFold(quoteType, quoteTypeEcdsa) && (akId.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP256 || akId.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP384) {
+		return attKeyId, nil
+	} else if (strings.EqualFold(quoteType, quoteTypeEpidUnlinkable) || strings.EqualFold(quoteType, quoteTypeEpidLinkable)) && akId.AlgorithmId == sgxQuoteLibraryAlgorithmEpid {
+		return attKeyId, nil
+	}
+
+	akIdNum, err := getSupportedAttKeyIDNumRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	akIdList, err := getSupportedAttKeyIDsRequest(akIdNum)
+	if err != nil {
+		return nil, err
+	}
+
+	var i uint32 = 0
+	for i < akIdNum {
+		attKeyId := akIdList[i*attestationKeyIdentityLength : (i+1)*attestationKeyIdentityLength]
+
+		akId := &attestationKeyIdentity{}
+		if err = restruct.Unpack(attKeyId, binary.LittleEndian, &akId); err != nil {
+			return nil, err
+		}
+
+		if strings.EqualFold(quoteType, quoteTypeEcdsa) && (akId.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP256 || akId.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP384) {
+			return attKeyId, nil
+		} else if (strings.EqualFold(quoteType, quoteTypeEpidUnlinkable) || strings.EqualFold(quoteType, quoteTypeEpidLinkable)) && akId.AlgorithmId == sgxQuoteLibraryAlgorithmEpid {
```

Comment:
ditto

Reply:
This is not a simple check but a traversal and conditional search.

Reply:
> ditto

This is not a simple check but a traversal and conditional search.